### PR TITLE
External lock logstore

### PR DESCRIPTION
**Pull request by batas**
The first version of new log store with external locking.

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/delta/storage/ExternalLockBaseLogStore.scala [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+import java.nio.file.FileSystemException
+
+import com.google.common.io.CountingOutputStream
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.delta.metering.DeltaLogging
+import org.apache.spark.sql.delta.util.FileNames
+
+abstract class ExternalLockBaseLogStore(sparkConf: SparkConf, hadoopConf: Configuration)
```
Comment (mrk-its) [inline src/main/scala/org/apache/spark/sql/delta/storage/ExternalLockBaseLogStore.scala]: I don't like this `ExternalLock` name part - what about ExternalKVLogStore? Or simply keep original name.

File: src/main/scala/org/apache/spark/sql/delta/storage/ExternalLockBaseLogStore.scala [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        logWarning(s"${e.getClass.getName}: ignoring recoverable error: $e")
+    }
+  }
+
+  /*
+   * Write cache in exclusive way.
+   * Method should throw @java.nio.file.FileAlreadyExistsException if path exists in cache.
+   *
+   */
+  protected def writeCacheExclusive(logEntry: LogEntryMetadata, fs: FileSystem): Unit
```
Comment (mrk-its) [inline src/main/scala/org/apache/spark/sql/delta/storage/ExternalLockBaseLogStore.scala]: I would keep single method with `overwrite` parameter (instead of two versions with and without Exclusive suffix) - it is closer to original LogStore interface, DynamoDBLogStore implementation also would be simpler.

Comment (mrk-its) [reply src/main/scala/org/apache/spark/sql/delta/storage/ExternalLockBaseLogStore.scala]: Also, I'm not quite convinced by naming convention (if these main abstract methods have "Cache" suffix maybe whole class also should be a "Cache"? But I don't like this idea). I would simply keep original naming.

File: src/main/scala/org/apache/spark/sql/delta/storage/MemoryLogStore.scala [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import com.google.common.cache.CacheBuilder
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileSystem, Path}
+import org.apache.spark.SparkConf
+
+import collection.JavaConverters._
+import scala.util.control.Breaks._
+
+
+class MemoryLogStore(sparkConf: SparkConf, hadoopConf: Configuration)
```
Comment (mrk-its) [inline src/main/scala/org/apache/spark/sql/delta/storage/MemoryLogStore.scala:30]: If this implementation is only for tests, it should be probably moved out of here (but then it will be problem with loading by test suite, right?)

File: src/main/scala/org/apache/spark/sql/delta/storage/ExternalLockBaseLogStore.scala [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+   * Returns path stripped user info.
+   */
+  protected def getPathKey(resolvedPath: Path): Path = {
+    stripUserInfo(resolvedPath)
+  }
+
+  /**
+   * Generate temporary path for TransactionLog.
+   */
+  protected def defGetTemporaryPath(path: Path): Path = {
```
Comment (mrk-its) [inline src/main/scala/org/apache/spark/sql/delta/storage/ExternalLockBaseLogStore.scala]: rename to `getTemporaryPath`?

File: src/test/scala/org/apache/spark/sql/delta/LogStoreSuite.scala [hunk 5]
```diff
@@ -197,10 +240,40 @@ class FakeAbstractFileSystem(uri: URI, conf: org.apache.hadoop.conf.Configuratio
     false) {
 
   // Implementation copied from RawLocalFs
+
```
Comment (mrk-its) [inline src/test/scala/org/apache/spark/sql/delta/LogStoreSuite.scala]: Let's avoid doing unrelated changes

File: src/main/scala/org/apache/spark/sql/delta/storage/MemoryLogStore.scala (hunk 3 above)
Comment (batas) [reply src/main/scala/org/apache/spark/sql/delta/storage/MemoryLogStore.scala:30]: Not exactly, this method should replace S3SingleDriverLogstore, because is almost the same implementation.

**Review by mmazek — APPROVED** [review]
